### DISCUSSION TITLE
Limitation in Filtering "ID" Parameter in runs.json API for Retrieving Specific Recipient Status

### DISCUSSION THREAD

**Discussion by ssweis** [nyaruka/rapidpro#4610] [Q&A]
We are currently in the process of developing a solution that involves utilizing APIs. However, we have encountered an issue with the "runs.json" API, specifically regarding the inability to effectively filter using the "ID" parameter, as it returns an empty result.

This limitation prevents us from retrieving the status of a specific recipient, which is a crucial requirement. Unfortunately, there doesn't seem to be an alternative solution to address this problem. Consequently, we are left with the only option of fetching the entire dataset for flow runs, disregarding the parameters of date and ID. This approach will have a performance impact, especially if the flow has been sent to a large number of recipients over a significant time period, as the dataset retrieved from the API will contain all the results of the flow.

Thank you for your attention to this matter.

**Comment by ericnewcomer** [comment]
A bit more detail here would be helpful. Can you describe the access pattern you are after in more detail? Are you polling for changes? What is driving the need to query by run id more specifically?

**Reply by rowanseymour** [reply]
The runs endpoint supports filtering by `id` or `uuid`.. I suspect this is filtering by the `start_id` which isn't possible but also.. maybe not needed.

**Reply by ericnewcomer** [reply]
Sorry, I was thinking start id in my head as well. I did also confirm that querying by run id and uuid both work, so we need more to go on here.

**Comment by ssweis** [comment]
![image](https://github.com/nyaruka/rapidpro/assets/134369207/8a1b6105-abf2-4026-b940-616bc135680f)
Is there a way to filter data using both the flow and the flow start ID? Currently, we can filter using the flow alone, but this retrieves all the flow data, which can be a large amount. 

In the screenshot below, you can see that we have applied a filter to the runs.json using only the flow parameter, resulting in the retrieval of all the data, including our flow starts from the previous steps.

Is it possible to filter the runs.json API using both the flow and the highlighted UUID in yellow from the previous two screenshots? Our objective is to obtain the specific data we need without having to filter the entire flow and manually search for the required UUID in the results.
![image](https://github.com/nyaruka/rapidpro/assets/134369207/4a9b862d-f3a4-4c5a-a928-11ba5075c23c)
Additionally, we can use the uuid filter in the run.json file, as shown in the screenshot below. However, the issue is that we don't receive the uuid in the flow_start API response during the first step. So, how can we obtain the uuid and pass it to the run API?
![image](https://github.com/nyaruka/rapidpro/assets/134369207/2c1af812-e2d4-4605-8057-557c4ddc9055)

**Reply by rowanseymour** [reply]
Ok so when you post to `flows_starts.json` you are creating a flow start object which is a thing that creates flow runs for multiple contacts in a background process. You might be creating millions of runs which takes a while so that has to happen asynchronously, and that endpoint can't return the ids/uuids of the runs themselves.

Can we take a step back here, and hear about what you're trying to do? Then we can advise on a better way to do this.

For example you could approach this differently and hit the `contacts.json` endpoint to create a contact from a phone number, e.g. `POST /api/v2/contacts.json?urn=tel%3A%2B250780000001`. Use that to get the UUID of the contact, and then use that to start flows and fetch runs (`runs.json` is filterable by contact UUID).
